spec/rubocop/cop/style/class_structure_spec.rb

```diff
@@ -0,0 +1,263 @@
+        validates :name
+
+        # public class methods are next in line
+        def self.some_method
+        end
+
+        # initialization goes between class methods and other instance methods
+        def initialize
+        end
+
+        # followed by other public instance methods
+        def some_method
+        end
+
+        # protected and private methods are grouped near the end
+        protected
+
+        def some_protected_method
+        end
+
+        private
+
+        def some_private_method
+        end
+      end
+    RUBY
+
+    it_behaves_like 'code without offense'
+  end
+  context 'simple example' do
+    let(:code) { <<-RUBY }
+      class Person
+        CONST = 'wrong place'
+        include AnotherModule
+        extend SomeModule
+      end
+    RUBY
+    let(:corrected_code) { <<-RUBY }
+      class Person
+        extend SomeModule
+        include AnotherModule
+        CONST = 'wrong place'
+      end
+    RUBY
+    let(:message) do
+      [
+        '`extend` is supposed to appear before `include` in the class definition.',
+        '`include` is supposed to appear before `constant` in the class definition.'
+      ]
+    end
+    it_behaves_like 'registers an offense'
+    it_behaves_like 'auto-corrects'
+  end
+
+  context 'move scopes with visibility' do
+    let(:code) { <<-RUBY }
+      class MyClass
+        def public_method
+        end
+
+        private
+
+        def first_private_method
+        end
+
+        def second_private_method
+        end
+
+        protected
+
+        def first_protected_method
+        end
+
+        def second_protected_method
+        end
+      end
+    RUBY
+    let(:corrected_code) { <<-RUBY }
+      class MyClass
+        def public_method
+        end
+
+        protected
+
+        def first_protected_method
+        end
+
+        def second_protected_method
+        end
+
+        private
+
+        def first_private_method
+        end
+
+        def second_private_method
+        end
+      end
+    RUBY
+    let(:message) { '`protected_method` is supposed to appear before `private_method` in the class definition.' }
+
+    it_behaves_like 'registers an offense'
+    it_behaves_like 'auto-corrects'
+  end
+  context 'complete example' do
+    let(:code) { <<-RUBY }
+      class Person
+        include AnotherModule
+        extend SomeModule
+
+        CustomError = Class.new(StandardError)
+
+        validates :name
+
+        attr_reader :name
+
+        def self.some_public_class_method
+        end
+
+        def initialize
+        end
+
+        def some_public_method
+        end
+
+        protected
+
+        def some_protected_method
+        end
+
+        private
+
+        def some_private_method
+        end
+      end
+    RUBY
+   let(:corrected_code) { <<-RUBY }
+      class Person
+        extend SomeModule
+        include AnotherModule
+        CustomError = Class.new(StandardError)
+        attr_reader :name
+
```

Comment:
I couldn't manage this extra line here. During the multiple swaps of the code, sometimes it happens. If anyone has any other better idea I'm really not sure if we need to take care. I think another cop will fix in in the loop.

I tried to pick the selection from the previous node end but also didn't work for all.

Reply:
If it adds lots of complexity, we can have the extra line cop pick it up in the correction loop.

Reply:
Ok! makes sense! Thanks, @Drenmi!